088-rebalance-progress-status.md

```diff
@@ -0,0 +1,392 @@
+However, the `executorState.json` information from when before the rebalance was interrupted or completed may still be of value to users.
+Therefore, we reuse the same value of `executorState.json` from the previous update.
+
+#### State: `Ready`
+
+The rebalance is complete in this state so there is no data surrounding ongoing partition movement available, so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+
+### Progress Update Cadence
+
+For ease of implementation and minimizing the load on the CruiseControl REST API server, the operator will only query the `/kafkacruisecontrol/state?substates=executor` endpoint and update the `ConfigMap` upon `KafkaRebalance` resource reconciliation.
+
+In the event that Cruise Control runs into an error when rebalancing, when operator transitions the `KafkaRebalance` resource to the `NotReady` state, it will leave the `progress` section and `ConfigMap` as is.
+In the event that the Cruise Control REST API returns an error or fails to respond to the operator when querying the `/kafkacruisecontrol/state?substates=executor` endpoint during a rebalance, the operator will add a "Warning" condition with error message from the failed REST API call but leave the existing progress section and progress `ConfigMap` as is.
+
+When Cruise Control state retrieval fails, the `KafkaRebalance` resource will be updated as follows:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  ...
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: Rebalancing
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: Warning
+    reason: CruiseControlRestException
+    message: <error_message> [1]  
+  observedGeneration: 1
+  optimizationResult:
+    afterBeforeLoadConfigMap: my-rebalance
+    dataToMoveMB: 0
+    excludedBrokersForLeadership: []
+    excludedBrokersForReplicaMove: []
+    excludedTopics: []
+    intraBrokerDataToMoveMB: 0
+    monitoredPartitionsPercentage: 100
+    numIntraBrokerReplicaMovements: 0
+    numLeaderMovements: 16
+    numReplicaMovements: 0
+    onDemandBalancednessScoreAfter: 95.4347095948149
+    onDemandBalancednessScoreBefore: 89.4347095948149
+    provisionRecommendation: ""
+    provisionStatus: RIGHT_SIZED
+    recentWindows: 1
+  progress:
+    rebalanceProgressConfigMap: my-rebalance
+```
+[1] Error message from failed Cruise Control REST API call.
+
+### Accessing progress fields using Kubernetes CLI
+
+The progress information will be stored in a `ConfigMap` with the same name as the `KafkaRebalance` resource.
+Using the name of the `ConfigMap` we can view its data from the command line using the Kubernetes CLI.
+
+Example accessing `estimatedTimeToCompletionInMinutes` field.
+```shell
+kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["estimatedTimeToCompletionInMinutes"]'
```

Comment:
What happens when you run this command and the field is "ommited" as you said above for some of the states. It will probably fail, but you should probably mention that.

Reply:
Fair point, added a note for this.